File: IEEEXTREME16/BookCipher/main.py
def findLetter(grid, c, pos):
    if pos == 'S':
        for k in range(len(grid)):
            for l in range(len(grid[k])):
                if grid[k][l] == c:
                    i = k
                    j = l
                    return [i + 1, j + 1]
    for k in range(len(grid)):
        for l in range(len(grid[k])):
            if grid[-k - 1][-l - 1] == c:
                i = len(grid) - k
                j = len(grid[k]) - l
                return [i, j]
    return [-1, -1]


# queue<string> phrases, XML;
chiper = []
phrases = []
XML = []
p = int(input())
n = int(input())
entry = input().split(",")
R = int(entry[0])
C = int(entry[1])
grid = [[]]
pos = input()
for i in range(p):
    phrases.append(input())
for i in range(n):
    XML.append(input())

i = j = 0
while i < R and j < C and len(XML) > 0:
    temp = XML.pop(0)
    k = 0
    while k < len(temp):
        if temp[k] == '<':
            while temp[k] != '>':
                k += 1
        else:
            grid[i].append(temp[k] if temp[k] != ' ' else '_')
            j += 1
            if j == C:
                j = 0
                i += 1
                if i == R:
                    break
                grid.append([])
        k += 1

positions = {}
while len(phrases) > 0:
    phrase = phrases.pop(0)
    chiper.append("")
    for c in phrase:
        if c not in positions:
            positions[c] = findLetter(grid, c, pos)
        if positions[c][0] == -1:
            chiper[-1] = "0 "
            break
        chiper[-1] += str(positions[c][0]) + "," + str(positions[c][1]) + ","
    chiper[-1] = chiper[-1][:-1]

for i in range(len(chiper)):
    print(chiper[i])

File: IEEEXTREME16/BookCipher/main.cpp
#include <bits/stdc++.h>
using namespace std;

pair<int, int> findLetter(const vector<vector<char>>& grid, char c, char pos) {
    int i, j;
    if (pos == 'S') {
        for (int k = 0; k < grid.size(); k++)
            for (int l = 0; l < grid[k].size(); l++)
                if (grid[k][l] == c) {
                    i = k;
                    j = l;
                    return make_pair(i + 1, j + 1);
                }
    }
    for (int k = grid.size() - 1; k >= 0; k--)
        for (int l = grid[k].size() - 1; l >= 0; l--)
            if (grid[k][l] == c) {
                i = k;
                j = l;
                return make_pair(i + 1, j + 1);
            }
    return make_pair(-1, -1);
}

int main() {
    int p, n, R, C;
    char pos;
    queue<string> phrases, XML;
    vector<string> chiper;
    cin >> p >> n;
    string temp;
    cin >> temp;
    R = stoi(temp.substr(0, temp.find(',')));
    C = stoi(temp.substr(temp.find(',') + 1));
    vector<vector<char>> grid(R, vector<char>(C));
    cin >> pos;
    for (int i = 0; i < p; i++) {
        cin >> temp;
        phrases.push(temp);
    }
    getline(cin, temp);
    for (int i = 0; i < n; i++) {
        getline(cin, temp);
        XML.push(temp);
    }

    int i = 0, j = 0;
    while (!XML.empty() && i < R && j < C) {
        temp = XML.front();
        XML.pop();
        for (int k = 0; k < temp.size(); k++) {
            if (temp[k] == '<')
                while (temp[k] != '>')
                    k++;
            else
                grid[i][j++] = temp[k] != ' ' ? temp[k] : '_';
            if (j == C) {
                j = 0;
                i++;
                if (i == R)
                    break;
            }
        }
    }

    unordered_map<char, pair<int, int>> positions;
    for (char c = 0; c < 127; c++)
        positions[c] = findLetter(grid, c, pos);

    while (!phrases.empty()) {
        temp = phrases.front();
        phrases.pop();
        chiper.emplace_back("");
        for (const char& c : temp) {
            if (positions.find(c) == positions.end())
                positions[c] = findLetter(grid, c, pos);
            if (positions[c].first == -1) {
                chiper.back() = "0 ";
                break;
            }
            chiper.back() += to_string(positions[c].first) + "," + to_string(positions[c].second) + ",";
        }
        chiper.back().pop_back();
    }

    for (const string& s : chiper)
        cout << s << endl;
/*
    // print positions
    for (const auto& p : positions)
        cout << p.first << " " << p.second.first << " " << p.second.second << endl;

    // print grid
    for (const auto& row : grid) {
        for (const auto& c : row)
            cout << c;
        cout << endl;
    }*/

    return 0;
}